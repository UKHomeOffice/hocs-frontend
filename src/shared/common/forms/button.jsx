import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { ApplicationConsumer } from '../../contexts/application.jsx';
import { cancel } from '../../contexts/actions/index.jsx';

class Button extends Component {

    handleClick(e, preventDefault) {
        if(preventDefault === 'true') {
            e.preventDefault();
        }

        this.props.dispatch(cancel());
    }

    render() {
        const {
            action,
            className,
            disabled,
            label,
            preventDefault
        } = this.props;
        return (
            <Fragment>
                <Link
                    className={`govuk-button${className ? ' ' + className : ''}`}
                    disabled={disabled}
                    to={action}
                    onClick={e => this.handleClick(e, preventDefault)}
                >{label}</Link>
            </Fragment>
        );
    }
}

Button.propTypes = {
    action: PropTypes.string.isRequired,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    dispatch: PropTypes.func.isRequired,
    label: PropTypes.string,
    preventDefault: PropTypes.string
};

Button.defaultProps = {
    disabled: false,
    preventDefault: 'true',
    label: 'Submit'
};

const WrappedButton = props => (
    <ApplicationConsumer>
        {({ dispatch }) => <Button {...props} dispatch={dispatch} />}
    </ApplicationConsumer>
);

export { WrappedButton, Button };
