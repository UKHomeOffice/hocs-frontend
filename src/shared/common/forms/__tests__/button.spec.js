import React from 'react';
import { Button } from '../button.jsx';
import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';

describe('Form button component', () => {

    const mockDispatch = jest.fn();

    beforeEach(() => {
        mockDispatch.mockReset();
    });

    it('should render with default props', () => {
        const wrapper = render(
            <Button action={'/SOME/URL'} label={'press'} dispatch={() => {}} />,
            { wrapper: MemoryRouter }
        );
        expect(wrapper).toBeDefined();
        expect(screen.getByText('press')).toBeInTheDocument();
    });

    it('should render disabled when isDisabled is passed', () => {
        const wrapper = render(
            <Button action={'/SOME/URL'} label={'press'} dispatch={() => {}} disabled={true} />,
            { wrapper: MemoryRouter }
        );
        expect(wrapper).toBeDefined();
        expect(screen.getByText('press')).toHaveAttribute('disabled');
    });

    it('should dispatch action when clicked', () => {
        global.testFunc = () => {};
        const testSpy = jest.spyOn(global, 'testFunc');
        const wrapper = render(
            <Button action={'/SOME/URL'} label={'press'} dispatch={testSpy} />,
            { wrapper: MemoryRouter }
        );

        expect(wrapper).toBeDefined();
        fireEvent.click(screen.getByText('press'));
        expect(testSpy).toHaveBeenCalledTimes(1);
    });

});
