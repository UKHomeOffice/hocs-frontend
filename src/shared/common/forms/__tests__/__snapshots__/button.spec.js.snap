// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form button component should render disabled when isDisabled is passed 1`] = `
<Button
  action="/SOME/URL"
  disabled={true}
  dispatch={[MockFunction]}
  label="Submit"
>
  <Link
    className="govuk-button"
    disabled={true}
    onClick={[Function]}
    to="/SOME/URL"
  >
    <LinkAnchor
      className="govuk-button"
      disabled={true}
      href="/SOME/URL"
      navigate={[Function]}
      onClick={[Function]}
    >
      <a
        className="govuk-button"
        disabled={true}
        href="/SOME/URL"
        onClick={[Function]}
      >
        Submit
      </a>
    </LinkAnchor>
  </Link>
</Button>
`;

exports[`Form button component should render with additional styles when className is passed 1`] = `
<Button
  action="/SOME/URL"
  className="test-class"
  disabled={false}
  dispatch={[MockFunction]}
  label="Submit"
>
  <Link
    className="govuk-button test-class"
    disabled={false}
    onClick={[Function]}
    to="/SOME/URL"
  >
    <LinkAnchor
      className="govuk-button test-class"
      disabled={false}
      href="/SOME/URL"
      navigate={[Function]}
      onClick={[Function]}
    >
      <a
        className="govuk-button test-class"
        disabled={false}
        href="/SOME/URL"
        onClick={[Function]}
      >
        Submit
      </a>
    </LinkAnchor>
  </Link>
</Button>
`;

exports[`Form button component should render with correct label when passed 1`] = `
<Button
  action="/SOME/URL"
  disabled={false}
  dispatch={[MockFunction]}
  label="My Button"
>
  <Link
    className="govuk-button"
    disabled={false}
    onClick={[Function]}
    to="/SOME/URL"
  >
    <LinkAnchor
      className="govuk-button"
      disabled={false}
      href="/SOME/URL"
      navigate={[Function]}
      onClick={[Function]}
    >
      <a
        className="govuk-button"
        disabled={false}
        href="/SOME/URL"
        onClick={[Function]}
      >
        My Button
      </a>
    </LinkAnchor>
  </Link>
</Button>
`;

exports[`Form button component should render with default props 1`] = `
<Button
  action="/SOME/URL"
  disabled={false}
  dispatch={[MockFunction]}
  label="Submit"
>
  <Link
    className="govuk-button"
    disabled={false}
    onClick={[Function]}
    to="/SOME/URL"
  >
    <LinkAnchor
      className="govuk-button"
      disabled={false}
      href="/SOME/URL"
      navigate={[Function]}
      onClick={[Function]}
    >
      <a
        className="govuk-button"
        disabled={false}
        href="/SOME/URL"
        onClick={[Function]}
      >
        Submit
      </a>
    </LinkAnchor>
  </Link>
</Button>
`;
